### PR TITLE
 Making DPR faster by avoiding unnecessary calculations

### PR DESCRIPTION
**Pull request by franzmpub**
# Pull Request      

Making DPR faster by avoiding unnecessary calculations

## Description

Since we don't need the conditional statement for "gold" passages starting on the original line 153:
```
               # if we have gold_pids, set their vector to 100 * the query vector
                         if ki < len(gpids):
                             doc_vectors[qi, ki] = 100 * query_vectors[qi

```
we don't need the calculation on line 162:

```
               doc_scores = torch.bmm(
                    query_vectors_tensor.unsqueeze(1), retrieved_doc_embeds.transpose(1, 2)
                ).squeeze(1)
```
Additionally, the `merge_results` function is moved outside `search`, to allow using it in `search_title_to_title`.
 
### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [x] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [x] Added [description](#description) of changes proposed.
- [x] Review requested as [appropriate](#request-review).
- [x] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: primeqa/ir/dense/dpr_top/dpr/searcher.py
Comment (seirasto) [inline primeqa/ir/dense/dpr_top/dpr/searcher.py:1]: Are the init_title_to_title and search_title_to_title functions used elsewhere? Can you add some documentation for them?

**Review by seirasto — APPROVED** [review]
LGTM. I've left a comment to add some description/documentation for the new functions.

**Review by bhavani105 — APPROVED** [review]
LGTM

**Review by hrojek — APPROVED** [review]
LGTM

File: primeqa/ir/dense/dpr_top/dpr/searcher.py
Comment (franzmpub) [reply primeqa/ir/dense/dpr_top/dpr/searcher.py:1]: These will be used in corresponding update to Components and Services.
The reason why they are part of this update is that `merge_results` is moved outside of `search`.

**Pull request merged**